python/mlx/nn/losses.py

```diff
@@ -4,163 +4,156 @@
-          ``'none'`` | ``'mean'`` | ``'sum'``. Default: ``'none'``.
+            ``'none'`` | ``'mean'`` | ``'sum'``. Default: ``'none'``.
 
     Returns:
-        mx.array: The computed cross entropy loss.
+        array: The computed cross entropy loss.
     """
+    if label_smoothing < 0 or label_smoothing >= 1:
+        raise ValueError(f"Label smoothing must in [0, 1), got {label_smoothing}.")
+
     score = mx.take_along_axis(logits, targets[..., None], axis).squeeze(-1)
-    loss = mx.logsumexp(logits, axis=axis) - score
+    logsumexp_logits = mx.logsumexp(logits, axis=axis)
+    if label_smoothing > 0:
+        # Adjust the true class score with label smoothing
+        adjusted_score = (1 - label_smoothing) * score
 
+        # Calculate the mean logit across the classes for smoothed loss
+        mean_logits = logits.mean(axis=axis)
+        smoothed_loss = -mean_logits * label_smoothing
+
+        # Combine the adjusted score and smoothed loss with the logsumexp logits
+        loss = logsumexp_logits - adjusted_score + smoothed_loss
+    else:
+        loss = logsumexp_logits - score
+
+    # Apply weights if provided
+    if weights is not None:
+        if weights.shape != targets.shape:
+            raise ValueError(
+                f"Weights with shape {weights.shape} is not the same as "
+                f"targets with shape {targets.shape}."
+            )
+        loss *= weights
+
+    # Apply reduction
     return _reduce(loss, reduction)
 
 
 def binary_cross_entropy(
-    inputs: mx.array, targets: mx.array, reduction: str = "none"
+    logits: mx.array, targets: mx.array, reduction: str = "none"
 ) -> mx.array:
     """
-    Computes the binary cross entropy loss between inputs and targets.
+    Computes the binary cross entropy loss.
 
     Args:
-        inputs (mx.array): The predicted inputs (post-sigmoid probabilities).
-        targets (mx.array): The target values (binary labels).
+        logits (array): The unnormalized (pre-sigmoid) predicted logits.
+        targets (array): The binary target values in {0, 1}.
         reduction (str, optional): Specifies the reduction to apply to the output:
           ``'none'`` | ``'mean'`` | ``'sum'``. Default: ``'none'``.
 
     Returns:
-        mx.array: The computed binary cross entropy loss.
+        array: The computed binary cross entropy loss.
     Examples:
         >>> import mlx.core as mx
         >>> import mlx.nn as nn
-        >>> inputs = mx.array([0.1, 0.2, 0.3, 0.4])
+        >>> inputs = mx.array([0.105361, 0.223144, 1.20397, 0.916291])
         >>> targets = mx.array([0, 0, 1, 1])
-        >>> loss = nn.losses.binary_cross_entropy(inputs, targets)
+        >>> loss = nn.losses.binary_cross_entropy(inputs, targets, "mean")
         >>> loss
-        array([0.612192])
+        array([0.612192], dtype=float32)
     """
-    loss = -targets * mx.log(inputs) - (1 - targets) * mx.log(1 - inputs)
+    loss = mx.logaddexp(0.0, logits) - targets * logits
```

Comment:
Why use `mx.logaddexp(0.0, logits)` for softmax and not just write it out to `log(1+exp(logits))`?